sw/airborne/firmwares/rotorcraft/autopilot.c

```diff
@@ -573,6 +572,77 @@ bool autopilot_guided_move_ned(float vx, float vy, float vz, float heading)
   return false;
 }
 
+/* Note: Offset position command in NED frame or body frame will only be implemented if
+ * local reference frame has been initialised. Additionally, if both the offset NED
+ * and offset body flags are set, the x,y setting is not implemented.
+ * Flag definition:
+   bit 0: x,y as offset coordinates
+   bit 1: x,y in body coordinates
+   bit 2: z as offset coordinates
+   bit 3: yaw as offset coordinates
+   bit 4: free
+   bit 5: x,y as vel
+   bit 6: z as vel
+   bit 7: yaw as rate
+ */
+void autopilot_guided_update(uint8_t flags, float x, float y, float z, float yaw)
+{
+  /* only update setpoints when in guided mode */
+  if (autopilot_mode != AP_MODE_GUIDED) {
+    return;
+  }
+
+  // handle x,y
+  if (bit_is_set(flags, 5)) { // velocity setpoint
+    if (bit_is_set(flags, 1)) { // set velocity in body frame
+      float psi = stateGetNedToBodyEulers_f()->psi;
+      x =  cosf(-psi) * x + sinf(-psi) * y;
+      y = -sinf(-psi) * x + cosf(-psi) * y;
+    }
+    guidance_h_set_guided_vel(x, y);
+  } else {  // position setpoint
+    if ((flags & 3) == 0) {   // set absolute position setpoint
```

Comment:
:( I wanted to check the two bits at the same time... ok, wilco

Kirk

On 01/06/16 13:32, Roland Meertens wrote:
>
> In sw/airborne/firmwares/rotorcraft/autopilot.c 
> <https://github.com/paparazzi/paparazzi/pull/1694#discussion_r65343865>:
>
> > +{
> > +  /* only update setpoints when in guided mode */
> > +  if (autopilot_mode != AP_MODE_GUIDED) {
> > +    return;
> > +  }
> > +
> > +  // handle x,y
> > +  if (bit_is_set(flags, 5)) { // velocity setpoint
> > +    if (bit_is_set(flags, 1)) { // set velocity in body frame
> > +      float psi = stateGetNedToBodyEulers_f()->psi;
> > +      x =  cosf(-psi) * x + sinf(-psi) * y;
> > +      y = -sinf(-psi) * x + cosf(-psi) * y;
> > +    }
> > +    guidance_h_set_guided_vel(x, y);
> > +  } else {  // position setpoint
> > +    if ((flags & 3) == 0) {   // set absolute position setpoint
>
> use bit_is_set
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/paparazzi/paparazzi/pull/1694/files/c64e3d95546f612c5d16ce11f85c54caea57da84#r65343865>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ACtEQ1EHhyipyiInPJf_FvBC94nl6C9aks5qHW26gaJpZM4IqAR->.
>